bin/rust/rust_belt/lifetime_logic.gh

```diff
@@ -126,23 +156,83 @@ Therefore the precondition does not need to be checked in the `requires` clause
 
 predicate frac_borrow(lifetime_t k, predicate(;) P);
 
+// LftL-bor-fracture
 lemma void full_borrow_into_frac(lifetime_t k, predicate(;) P);
+    nonghost_callers_only
     requires full_borrow(k, P);
     ensures [_]frac_borrow(k, P);
 
+// LftL-fract-shorten
 lemma void frac_borrow_mono(lifetime_t k, lifetime_t k1, predicate(;) P);
     requires [_]frac_borrow(k, P) &*& lifetime_inclusion(k1, k) == true;
     ensures [_]frac_borrow(k1, P);
 
 predicate close_frac_borrow_token(real q_p, predicate(;) P, real q_lft, lifetime_t k);
 
-// Rule LftL-fract-acc
+// LftL-fract-acc
 lemma void open_frac_borrow(lifetime_t k, predicate(;) P, real q_lft);
+    nonghost_callers_only
     requires [_]frac_borrow(k, P) &*& [q_lft]lifetime_token(k);
     ensures [?q_p]P() &*& close_frac_borrow_token(q_p, P, q_lft, k);
 
 lemma void close_frac_borrow(real q_p, predicate(;) P);
+    nonghost_callers_only
     requires [q_p]P() &*& close_frac_borrow_token(q_p, P, ?q_lft, ?k);
     ensures [q_lft]lifetime_token(k);
 
+// Mask preserving view shifts - mask aware versions. See the documentation about Iris view shifts in general.h
+// LftL-begin
+lemma lifetime_t begin_lifetime_m();
+    requires atomic_mask(Nlft);
+    ensures atomic_mask(Nlft) &*& lifetime_token(result);
+
+lemma void end_lifetime_m(lifetime_t k);
+    requires atomic_mask(Nlft) &*& lifetime_token(k);
+    ensures atomic_mask(Nlft) &*& [_]lifetime_dead_token(k);
+
+// LftL-bor-acc
+lemma void open_full_borrow_m(real q, lifetime_t k, predicate() P);
+    requires atomic_mask(Nlft) &*& full_borrow(k, P) &*& [q]lifetime_token(k);
+    ensures atomic_mask(Nlft) &*& P() &*& close_full_borrow_token(P, q, k);
+
+lemma void close_full_borrow_m(predicate() P);
+    requires atomic_mask(Nlft) &*& close_full_borrow_token(P, ?q, ?k) &*& P();
+    ensures atomic_mask(Nlft) &*& full_borrow(k, P) &*& [q]lifetime_token(k);
+
+lemma void reborrow_m(lifetime_t k1, lifetime_t k, predicate() P);
+    requires atomic_mask(Nlft) &*& lifetime_inclusion(k1, k) == true &*& full_borrow(k, P);
+    ensures atomic_mask(Nlft) &*& full_borrow(k1, P) &*& end_reborrow_token(k1, k, P);
+
+lemma void end_reborrow_m(lifetime_t k1, lifetime_t k, predicate() P);
+    requires atomic_mask(Nlft) &*& end_reborrow_token(k1, k, P) &*& [_]lifetime_dead_token(k1);
+    ensures atomic_mask(Nlft) &*& full_borrow(k, P);
+
+// LftL-bor-na
+lemma void full_borrow_into_nonatomic_borrow_m(lifetime_t k, thread_id_t t, mask_t m, predicate() P);
+    requires atomic_mask(Nlft) &*& full_borrow(k, P) &*& !mask_is_empty(m);
+    ensures atomic_mask(Nlft) &*& [_]nonatomic_borrow(k, t, m, P);
+
+// LftL-na-acc
+lemma void open_nonatomic_borrow_m(lifetime_t k, thread_id_t t, mask_t m, real q);
+    requires atomic_mask(mask_union(Nlft, m)) &*& [_]nonatomic_borrow(k, t, m, ?P) &*& [q]lifetime_token(k) &*& partial_thread_token(t, m);
```

Comment:
Is it necessary to require `m` in the atomic mask here?

Reply:
LftL-na-acc demands it
[https://gitlab.mpi-sws.org/iris/lambda-rust/-/blob/master/lifetime/na_borrow.v?ref_type=heads#L39](url)

Reply:
OK, then yes, let's require it here, too. I guess it doesn't matter too much anyway, since nonatomic borrows are probably typically opened, well, nonatomically, so with top mask available.